### PR TITLE
Bot Fixed Point and Moore Machine Servers

### PR DESCRIPTION
**Pull request by solomon-b**
This is @masaeedu's Bot Fixed Point PR with merge conflicts from main resolved. I made a PR against' his fork with this fix but he is too busy and suggested I just make a new PR.  I'll attempt to summarize the changes here.

We now have the `Behavior` type which represents the fixed point of a `Bot`:
```haskell
newtype Behavior m i o = Behavior { runBehavior :: i -> m (o, (Behavior m i o)) }
```

Notice that the `s` has disappeared. This should allow us to build up bots with complex state without ever growing `s` types in our Bot signatures which in turn would help with composability. 

We construct a `Behavior` from a `Bot` with `fixBot`:
```haskell
fixBot :: Functor m => Bot m s i o -> s -> Behavior m i o
fixBot (Bot b) = go
  where
  go s = Behavior $ \i -> second go <$> b s i
```

Next we introduce `Env` and `Server` which are sort of the dual to a `Bot` and `Behavior`:
```haskell
type Env :: KBot
newtype Env m s o i = Env { runEnv :: s -> (i, o -> m s) }

newtype Server m o i = Server { runServer :: (i, o -> m (Server m o i)) }

fixEnv :: Functor m => Env m s o i -> s -> Server m o i
fixEnv (Env b) = go
  where
  go s = Server $ fmap (fmap (fmap go)) $ b s
```

```haskell
newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
newtype Env m s o i = Env { runEnv :: s -> (i, o -> m s) }

newtype Behavior m i o = Behavior { runBehavior :: i -> m (o, (Behavior m i o)) }
newtype Server   m o i = Server   { runServer   :: (i, o -> m (Server m o i )) }
```

Notice that the `Bot` recieves an `i` and produces an `o` whereas the `Server` receives an `o` and produces an `i`.

The chat bot `Behavior` is then paired with a `Server` to produce a `Fix m` which we recursively loop over to run our bot:
```haskell
annihilate :: Monad m => Behavior m i o -> Server m o i -> Fix m
annihilate (Behavior a) (Server (i, b)) = Fix $ do
  (o, a') <- a i
  b' <- b o
  return $ annihilate a' b'

loop :: Monad m => Fix m -> m x
loop (Fix x) = x >>= loop
```

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -13,6 +13,9 @@ import qualified Options.Applicative           as Opt
 import           OptionsParser
 import           System.Environment.XDG.BaseDir ( getUserCacheDir )
 import           System.Process.Typed
+import           Control.Monad.Except (ExceptT, runExceptT)
+import Control.Monad.IO.Class (liftIO)
```
Comment (isovector) [inline app/Main.hs]: nit: indent

File: src/CofreeBot/Bot.hs [hunk 2]
```diff
@@ -135,26 +182,41 @@ mapMaybeBot f (Bot bot) =
...
+
+  -- Log in
+  userId   <- liftMatrixIO $ getTokenOwner session
+  filterId <- liftMatrixIO $ createFilter session userId messageFilter
+
+  -- Start looping
+  go filterId since
+
+  where
+    go filterId since = do
```
Comment (isovector) [inline src/CofreeBot/Bot.hs]: type def would be nice

File: app/Main.hs [hunk 3]
```diff
@@ -51,10 +54,15 @@ cliMain :: IO ()
 cliMain = withProcessWait_ ghciConfig $ \process -> do
   void $ threadDelay 1e6
   void $ hGetOutput (getStdout process)
-  runTextBot (simplifyMatrixBot $ bot process) mempty
+  repl' <- repl
+  loop $ annihilate (flip fixBot mempty $ simplifyMatrixBot $ bot process) repl'
+
+absorbError :: Show e => ExceptT e IO a -> IO a
+absorbError = runExceptT >=> either (fail . show) pure
```
Comment (isovector) [inline app/Main.hs]:
```suggestion
unsafeCrashInIO = runExceptT >=> either (fail . show) pure
```

File: src/CofreeBot/Bot.hs [hunk 4]
```diff
@@ -1,3 +1,7 @@
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# HLINT ignore "Use <$>" #-}
+{-# LANGUAGE RankNTypes #-}
```
Comment (isovector) [inline src/CofreeBot/Bot.hs:1]: nit: sort these

File: app/Main.hs [hunk 5]
```diff
@@ -51,10 +54,15 @@ cliMain :: IO ()
...
 
 matrixMain :: ClientSession  -> String -> IO ()
 matrixMain session xdgCache = withProcessWait_ ghciConfig $ \process -> do
   void $ threadDelay 1e6
   void $ hGetOutput (getStdout process)
-  runMatrixBot session xdgCache (bot process) mempty
+  matrix' <- absorbError $ matrix session xdgCache
+  absorbError $ loop $ annihilate (fmap join $ traverse' $ flip fixBot mempty $ hoistBot liftIO $ bot process) matrix'
```
Comment (isovector) [inline app/Main.hs]: easier to read:

```suggestion
  absorbError 
    $ loop 
    $ flip annihilate matrix' 
    $ fmap join 
    $ traverse' 
    $ flip fixBot mempty 
    $ hoistBot liftIO 
    $ bot process 
```

Comment (isovector) [reply app/Main.hs]: but also wat

File: src/CofreeBot/Bot.hs [hunk 6]
```diff
@@ -47,6 +53,48 @@ newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
     (Functor, Applicative, Monad, MonadState s, MonadReader i, MonadIO)
   via StateT s (ReaderT i m)
 
+newtype Behavior m i o = Behavior { runBehavior :: i -> m (o, (Behavior m i o)) }
```
Comment (isovector) [inline src/CofreeBot/Bot.hs:59]: https://hackage.haskell.org/package/machines-0.7.3/docs/Data-Machine-MealyT.html#t:MealyT

File: src/CofreeBot/Bot.hs [hunk 7]
```diff
@@ -47,6 +53,48 @@ newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
...
+    $ flip runStateT b
+    $ traverse (\i -> StateT $ \(Behavior b') -> fmap (\(responses, nextState) -> (responses, nextState)) $ b' i) is
+
+fixBot :: Functor m => Bot m s i o -> s -> Behavior m i o
+fixBot (Bot b) = go
+  where
+  go s = Behavior $ \i -> second go <$> b s i
+
+type Env :: KBot
+newtype Env m s o i = Env { runEnv :: s -> (i, o -> m s) }
```
Comment (isovector) [inline src/CofreeBot/Bot.hs]: Are `i` and `o` backwards here? Also this thing can and should have a functor and profunctor instance.

File: src/CofreeBot/Bot.hs [hunk 8]
```diff
@@ -47,6 +53,48 @@ newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
...
+
+newtype Server m o i = Server { runServer :: (i, o -> m (Server m o i)) }
```
Comment (isovector) [inline src/CofreeBot/Bot.hs]: https://hackage.haskell.org/package/machines-0.7.3/docs/Data-Machine-MooreT.html#t:MooreT, if you're willing to delay the first output behind a monadic action

File: src/CofreeBot/Bot.hs [hunk 9]
```diff
@@ -47,6 +53,48 @@ newtype Bot m s i o = Bot { runBot :: s -> i -> m (o, s) }
...
+instance Monad m => Traversing (Behavior m)
+  where
+  -- TODO: write wander instead for efficiency
+
+  traverse' b = Behavior $ \is ->
+    fmap (uncurry (,) . fmap traverse')
+    $ flip runStateT b
+    $ traverse (\i -> StateT $ \(Behavior b') -> fmap (\(responses, nextState) -> (responses, nextState)) $ b' i) is
+
+fixBot :: Functor m => Bot m s i o -> s -> Behavior m i o
```
Comment (isovector) [inline src/CofreeBot/Bot.hs]: imo this would be more satisfying as `Behavior m (i, s) (o, s) -> Behavior m i o`, at which point it's just costrength

**Review by isovector — COMMENTED** [review]
Left some comments, but they're all below the level of understanding the high-level picture here. At a higher level, inline documentation would be :+1:. There are lots of types that just show up and it's unclear why or what they're doing.

File: src/CofreeBot/Bot.hs (hunk 7 above)
Comment (solomon-b) [reply src/CofreeBot/Bot.hs]: I believe this is correct as the `Env` receives an output from a `Bot` and produces the next input to the `Bot`.

**Comment by solomon-b** [comment]
> Left some comments, but they're all below the level of understanding the high-level picture here. At a higher level, inline documentation would be +1. There are lots of types that just show up and it's unclear why or what they're doing.

Thanks so much for this feedback. I've added some haddocks but I feel unable to write documentation for the fixed point stuff. Its the core of this PR but I still have a lot of trouble understanding how the state threading works once you are in the fixed point of a `Bot` and `Env`. @masaeedu if you have any time could you try to summarize how this works for me? I can then turn that into documentation.

EDIT: Actually it just sorta clicked how it is handled in `fixEnv` and `fixBot`. I'll carry on with more haddocks.

**Review by isovector — APPROVED** [review]

**Pull request merged**